docs/DSP/GameCube_DSP_Users_Manual/GameCube_DSP_Users_Manual.tex

```diff
@@ -4338,9 +4355,18 @@ \section{Extended opcodes}
+(\Opcode{NX} has no behavior of its own, so it can be used to get an extended opcode's behavior on its own.)
 
 Extended opcodes do not modify the program counter (\Register{\$pc} register).
 
+Extended opcodes are run \textit{in parallel} with the main opcode; they see the same register state as the input. (For instance, \texttt{\Opcode{MOVR}\Opcode{'MV} \Register{\$ac1}, \Register{\$ax0.l} : \Register{\$ax0.l}, \Register{\$ac1.m}} (encoded as \Value{0x6113}) \textit{swaps} the values of \Register{\$ac1.m} and \Register{\$ax0.l} (and also extends the new value of \Register{\$ac1.m} into \Register{\$ac1.l} and \Register{\$ac1.h}).)
+
+Since they are executed in parallel, the main and extension opcodes could theoretically write to the same registers. All opcodes that support extension only modify a main accumulator \Register{\$acD}, as well as \Register{\$prod}, \Register{\$sr}, and/or \Register{\$pc}, while the extension opcodes themselves generally only modify an additional accumulator \Register{\$axD} and addressing registers \Register{\$arS}. The exception is \Opcode{'L} and \Opcode{'LN}, which has the option of writing to \Register{\$acD}. Thus, \texttt{\Opcode{INC}\Opcode{'L} \Register{\$ac0} : \Register{\$ac0.l}, @\Register{\$ar0}} (encoded as \Value{0x7660}) increments \Register{\$ac0} (and thus \Register{\$ac0.l}), but also sets \Register{\$ac0.l} to the value in data memory at address \Register{\$ar0} and increments \Register{\$ar0}.
+
+When the main and extension opcodes write to the same register, the register is set to the two values bitwise-or'd together. For the above example, \Register{\$ar0.l} would be set to \InlineExpression{(\Register{\$ar0.l} + 1) | MEM[\Register{\$ar0}]}. \textbf{Note that no official uCode writes to the same register twice like this.}
+
+Dolphin only supports this multi-write behavior in the DSP Interpreter when the \texttt{PRECISE\_BACKLOG} define is present. In \texttt{Interpreter::ExecuteInstruction}, if an extended opcode is in use, the extended opcode's behavior is executed first, followed by the main opcode's behavior. The extended opcode does not directly write to registers, but instead records the writes into a backlog (\texttt{WriteToBackLog}). The main opcode calls \texttt{ZeroWriteBackLog} after it is done reading the register values; this directly writes zero to all registers that have pending writes in the backlog. The main opcode then is free to write directly to registers it changes. Afterwards, \texttt{ApplyWriteBackLog} bitwise-ors the value of the register and the value in the backlog; if the main opcode didn't write to the register then \texttt{ZeroWriteBackLog} means that the pending value is being or'd with zero, so it's used without changes. When \texttt{PRECISE\_BACKLOG} is not defined, \texttt{ZeroWriteBackLog} does nothing and \texttt{ApplyWriteBackLog} overwrites the register value with the value from the backlog (so writes from extended opcodes ``win'' over the main opcode).
```

Comment:
I don't think we should be including Dolphin specific implementation information in a spec document that's intended to document only the hardware. This should be in the source code somewhere rather than here.

Reply:
I kinda agree, but I feel like it's important to at least mention that Dolphin doesn't emulate it by default and how to enable it, and if that's done explaining how it's implemented is somewhat helpful.

Most of this information is already present in the source code (`WriteToBackLog` in `DSPInterpreter.h` and `ApplyWriteBackLog` and `ZeroWriteBackLog` in `DSPInterperter.cpp`), just without a higher-level overview.

Reply:
I still think documentation to notify people of emulation differences shouldn't be in this document because most people are not going to read hardware documentation to find out about a software difference.

Maybe consider making a top-level `QUIRKS.md` or some similar document in the source tree dedicated for documenting emulation differences (for both intentional and unintentional choices), which gives it way more visibility than in a latex document tucked away inside the `docs/DSP` folder

It also easier to maintain because most people are not going to think about updating the DSP document if things change in the future

Reply:
I'm going to wait for a second opinion on this (while I work on some other DSP-related stuff).

Reply:
I agree with lioncash.

Reply:
> I still think documentation to notify people of emulation differences shouldn't be in this document because most people are not going to read hardware documentation to find out about a software difference.

My reasoning is that if someone were to see something like `INC’L $ac0 : $ac0.l, @$ar0` in uCode that Dolphin seems to not be handling correctly, it would make sense to check the hardware documentation to determine what's supposed to happen in that case, and the explanation of where and how it's implemented in Dolphin would be useful in that case.

I don't think something like `QUIRKS.md` is a bad idea, but I'm not up for writing something like that right now.  I guess I'll just remove this paragraph for the time being.

Reply:
> I don't think something like QUIRKS.md is a bad idea, but I'm not up for writing something like that right now. I guess I'll just remove this paragraph for the time being.

Sounds good

Reply:
I've moved the paragraph next to the (commented out) `PRECISE_BACKLOG` define, and also moved a few comments so that they're all in DSPInterpreter.cpp (instead of being mixed between that and DSPInterperter.h).